harness/determined/estimator/_reducer.py

```diff
@@ -0,0 +1,138 @@
+import abc
+from typing import Any, Callable, List
+
+import numpy as np
+import tensorflow as tf
+
+from determined import estimator
+
+
+class MetricReducer:
+    """
+    Efficiently aggregating metrics across a multi-slot distributed evaluation is done in two steps:
+
+      1. Accumulate metrics from each batch on each slot. In the case of calculating a mean, this
+         might mean keeping a running sum and a count of metrics received.
+
+      2. Reduce metrics from each slot to calculate the final metric. In the case of calculating a
+         mean, this might mean adding up the per-slot sums and dividing the result by the per-slot
+         counts.
+
+    Usage:
+
+        def my_estimator_model_function(features, labels, mode):
+            ...
+            if mode == tf.estimator.ModeKeys.EVAL:
+
+                my_avg_prediction = context.make_metric(predictions, MyAvgMetricReducer())
+
+                return tf.estimator.EstimatorSpec(
+                    mode,
+                    loss=loss,
+                    eval_metric_ops={"my_avg_prediction": my_avg_prediction},
+                )
+    """
+
+    @abc.abstractmethod
+    def reset(self) -> None:
+        # TODO: wait! when and how is this ever called?
+        pass
+
+    @abc.abstractmethod
+    def accumulate(self, metric: Any) -> Any:
+        """
+        accumulate is called for each batch in the evaluation dataset.  Batches will be distributed
+        across slots, so accumulate will be called many times on each slot.
+
+        accumulate should return the accumulated state.  After evaluation is complete, the final
+        return value of accumulate will become an element of the per_slot_metrics argument to
+        cross_slot_reduce.
+
+        In the example of the calculating a distributed mean, accumulate might keep a running sum
+        and a count of metrics received:
+
+            def accumulate(self, metric):
```

Comment:
praise: really nice informative doc strings

Reply:
thank you!